### PR TITLE
Implementation of cover driver for Buchhandel.de

### PR DESCRIPTION
**Pull request by jochen-lienhard**
Buchhandel.de supports covers, you need a token to get access to the covers.

### CODE REVIEW

**Comment by jochen-lienhard** [comment]
The images must be linked to the link-url in the config including the isbn. Do you have any idea, how we can manage this in the cover.phtml?
Next days I will implement the test-case.

**Comment by demiankatz** [comment]
The problem here is that the current mechanism is entirely based on image loading -- we just include an image tag that points to the cover, and then the cover loader sends back an image. There's no way of communicating additional information (like the need for an extra link) using this mechanism.

In order to achieve this sort of effect, I think we would need to develop a completely different mechanism for cover loading -- for example, instead of proxying the image through the cover loader as we currently do, we could set up an AJAX endpoint that sends back information about the image that should be loaded, and inject it directly into the browser using Javascript. This would allow us to pass along extra details such as required links.

Obviously, the AJAX approach has some disadvantages (greater complexity than the simple "just load an image" strategy, a dependency on Javascript, etc.).

I could also imagine a scenario where there might be some desire to combine both mechanisms -- i.e. try to load some services using the standard cover mechanism, then fail over to Javascript if that doesn't work. I'm not sure if such a thing is actually possible, however.

Bottom line: it would obviously be useful to be able to insert links in order to meet requirements of APIs like this one (and Amazon, for that matter). However, that's going to require significant architecture work, and if we do that work, it's not yet clear to me whether/how we need that to coexist with, or completely replace, the existing mechanisms.

I'll talk to Chris and see if he has any thoughts -- I'd also be interested to hear yours. Maybe we can start work on some kind of proof of concept for discussion at a future dev call; I'm pretty sure that others share this need.

**Comment by EreMaijala** [comment]
The upside of direct image loading is easy caching. If image links are added via AJAX, they can't be loaded quite as quickly as bare images. Having that as an option would be useful for sure, but let's not replace the simple image loading.

**Comment by demiankatz** [comment]
Yes, we should definitely maintain the simple image loading (if for no other reason than there are some uses that cannot be solved in any other way -- e.g. loading images from the file system based on patterns). Just a question of figuring out how to allow the two options to coexist most sensibly.

**Comment by lahmann** [comment]
Probably another thing to consider is icon-support as an alternative to covers: right now we have a fallback solution for non-existing covers with a static image, (optionally) falling back to icons that are e.g. chosen by record-format could be a useful addition. What do you think?

**Comment by demiankatz** [comment]
VuFind already supports format-based icons through the cover loader -- this mechanism is currently used in the Summon module.

**Comment by lahmann** [comment]
Oh, that's interesting, I didn't know that.

**Comment by jochen-lienhard** [comment]
I think the easiest way is to modify the cover.phtml. I'm not sure if we should add this to this PR.

**Comment by demiankatz** [comment]
Yes, it makes sense to put the logic into cover.phtml to trigger whatever AJAX routine needs to occur... but all that AJAX plumbing needs to be built. Also, we should think about where the link needs to go and how it should be formatted. Will it always need to be next to the thumbnail, or might there be some situations where we just need to add a "powered by ..." note at the bottom of the page, or something similar?

**Comment by jochen-lienhard** [comment]
In our current implementation (vufind 1.3) the thumbnail itself has the link. There problem is, that there is not only one way to add such link. I don't know the API of the other cover provider, but I think for amazon you must add a link for "here you can buy", like it is necessary for buchhandel.de too.

**Comment by jochen-lienhard** [comment]
As a dirty fix you could add something like this to the getCover method in the Record.php Root-Helper

       if (isset($this->config->Buchhandel->link)) {
            $isbn = is_callable([$this->driver, 'getCleanISBN'])
            ? $this->driver->getCleanISBN() : '';
            $link=$this->config->Buchhandel->link . $isbn;
        }

Perhaps we can modify it and add a 'link' paramater to the Content.

**Comment by demiankatz** [comment]
The problem with the "dirty fix" is that it will display the link whether or not a Buchhandel cover is provided -- since the API presumably will not have an image for every ISBN in existence. If you have a series of different image providers configured, you need to know which provider actually provides an image in order to know which message to display. That's where it gets complicated, because the template code doesn't know which provider will be used, because that selection happens asynchronously as part of the image loading process. However, this workaround is a simple solution if you only use one provider and don't mind occasionally displaying links below "missing" images.

**Comment by jochen-lienhard** [comment]
The link will only be set, if a cover is available. 
Your are right with the problem, if you use different cover provider. 
I think the problem of add a link to the provider (by cover or by separate icon or text), should not be solved in this PR. I only want to add the base function to show bookcover from buchhandle.de in this PR. The linking problem is a more common problem of the covers, so we should open a new PR for that.

**Comment by demiankatz** [comment]
I still think there are probably some edge cases where a link might be set for an unavailable cover -- at the time the template renders the link, we can only determine whether we have parameters that might return us a cover -- we haven't yet checked whether those parameters ACTUALLY retrieve a cover. But if the collection of cover images is very large, the set of records that will have this problem is hopefully fairly small (i.e. records with invalid ISBNs in the MARC, etc.).

In any case, I agree that we can get this cover provider merged as a first step (as long as we put good notes in the .ini file about complying with terms of use, since that will require some manual effort on the part of the user). We can deal with the bigger problem separately.

I'll try to review this more closely later in the week.

**Comment by jochen-lienhard** [comment]
I'm not sure, if we get always the correct ISBN. For Buchhandel.de you need the ISBN13.

**Comment by demiankatz** [comment]
ISBN10's shouldn't cause a problem, because the Buchhandel handler receives an ISBN object, and the object will automatically convert ISBN10 to ISBN13 when you call the get13() method.

**Comment by demiankatz** [comment]
I just pushed up some minor changes intended to get the Travis build to pass. When you get a chance, please give it one more test and let me know if I've broken anything (though I expect it should be fine). Also, note that #514 was just opened with another implementation of the same API. If you have a chance, please take a look and see if anything there should be adapted into this version.

**Comment by jochen-lienhard** [comment]
It looks good. I'm not sure about the url-config. How do you handle it, should url's end with a slash or should we add the slash to the return value?
I get a look to the other PR, but I found no further feature, I will add a comment there.
Shall we add more comment to the config part (for example that Buchhandel does not allow to cache the covers, and that a link is neccessary)?

**Comment by demiankatz** [comment]
Regarding slashes in the config, I'd recommend that we do a `trim($configSetting, '/')` and then add the slash back on somewhere in the code. This way, it will work whether or not the configuration includes a slash. As for comments, I agree that it would be a good idea to add a comment about adding links, etc. If they have a URL for their terms of service, a link might be helpful too.

**Comment by jochen-lienhard** [comment]
Do you need more modifications?

**Comment by demiankatz** [comment]
I made a few adjustments to the comments in config.ini and removed the link setting (since it's not being used yet). I've merged it all to master. Thanks again for sharing this!

**Comment by demiankatz** [comment]
Here's a JIRA ticket for continuing discussion of AJAX image information loading: https://vufind.org/jira/browse/VUFIND-1140
